yabt/builders/cpp.py

```diff
@@ -21,38 +21,169 @@
+        build_context, target, entrypoint=entrypoint)
+
+
+register_builder_sig(
+    'CppProg',
+    [('binary', PT.File),
+     ('sources', PT.FileList),
+     ('in_buildenv', PT.Target),
+     ('headers', PT.FileList, None),
+     ('cmd_env', None),
+     # ('copy_bin_to', PT.File, None),
+     ])
+
+
+@register_manipulate_target_hook('CppProg')
+def cpp_prog_manipulate_target(build_context, target):
+    target.buildenv = target.props.in_buildenv
+
+
+def compile_cc(build_context, buildenv, sources, workspace_dir,
+               buildenv_workspace, cmd_env):
+    """Compile list of C++ source files in a buildenv image
+       and return list of generated object file.
+    """
+    objects = []
+    for src in sources:
+        obj_rel_path = '{}.o'.format(splitext(src)[0])
+        obj_file = join(buildenv_workspace, obj_rel_path)
+        # TODO: compiler flags should come from target & project config
+        obj_cmd = [
+            CC, '-o', obj_file, '-c',
+            '-std=c++11', '-Wall', '-fvectorize', '-fslp-vectorize',
+            '-fcolor-diagnostics', '-O2', '-DDEBUG',
+            '-I{}'.format(buildenv_workspace),
+            join(buildenv_workspace, src)]
+        build_context.run_in_buildenv(buildenv, obj_cmd, cmd_env)
+        objects.append(
+            join(relpath(workspace_dir, build_context.conf.project_root),
+                 obj_rel_path))
+    return objects
+
+
+# TODO: make configurable
+CC = 'clang++-5.0'
+
+
+@register_build_func('CppProg')
+def cpp_prog_builder(build_context, target):
+    """Build a C++ binary executable"""
+    yprint(build_context.conf, 'Build CppProg', target)
+    workspace_dir = build_context.get_workspace('CppProg', target.name)
+    all_files = target.props.sources + target.props.headers
+    # add headers of direct dependencies
+    for dep in build_context.generate_direct_deps(target):
+        all_files.extend(dep.props.get('headers', []))
+    # add objects of all dependencies (direct & transitive)
+    objects = []
+    for dep in build_context.generate_all_deps(target):
+        objects.extend(dep.props.get('objects', []))
+    all_files.extend(objects)
+    link_artifacts(all_files, workspace_dir, None, build_context.conf)
+    buildenv_workspace = build_context.conf.host_to_buildenv_path(
+        workspace_dir)
+    objects.extend(compile_cc(
+        build_context, target.props.in_buildenv, target.props.sources,
+        workspace_dir, buildenv_workspace, target.props.cmd_env))
+    bin_file = join(buildenv_workspace, target.props.binary)
+    # TODO: linker flags should come from target & project conf
+    bin_cmd = [CC, '-o', bin_file] + objects
+    build_context.run_in_buildenv(
+        target.props.in_buildenv, bin_cmd, target.props.cmd_env)
+    target.artifacts['gen'] = {
+        join('bin', *split(target.name)):
+        join(workspace_dir, *split(target.name))
+    }
+
+    # # Copy binary artifacts to external destination
```

Comment:
why not build it there ?

Reply:
it's not a hermetic environment. I rather not even support the "copy-to-bin-dir" option.

Reply:
We agreed that this is overly 20 century of me